docs/src/docs/asciidoc/end-to-end-gradle-guide.adoc

```diff
@@ -0,0 +1,871 @@
+      },
+      ...
+    ],
+    "bundles": [],
+    "jni": [
+      {
+        "condition": {
+          "typeReached": "java.net.InetAddress"
+        },
+        "type": "java.lang.Boolean",
+        "methods": [
+          {
+            "name": "getBoolean",
+            "parameterTypes": [
+              "java.lang.String"
+            ]
+          }
+        ]
+      }
+    ]
+  }
+----
+
+As you can see, there are lots of entries that you may don't want.
+To reduce the amount of generated metadata, we will use the following `user-code-filter.json`:
+
+[source,json,subs="verbatim,attributes", role="multi-language-sample"]
+----
+{
+  "rules": [
+    {"includeClasses": "**"},
+    {"excludeClasses": "org.junit.**"},
+    {"excludeClasses": "org.gradle.**"},
+    {"excludeClasses": "worker.org.gradle.**"},
+    {"excludeClasses": "org.slf4j.**"},
+    {"excludeClasses": "java.**"}
+  ]
+}
+----
+
+[WARNING]
+====
+Be careful to remove only redundant metadata and keep metadata necessary for your project.
+====
+
+After we regenerate the metadata with the new filter, `reachability-metadata.json` generated on the same example as above will look like this:
+
+[source,json,subs="verbatim,attributes", role="multi-language-sample"]
+----
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "org.example.NativeTests"
+      },
+      "type": "org.example.NativeTests$Person",
+      "allDeclaredFields": true
+    },
+    {
+      "condition": {
+        "typeReached": "sun.security.jca.GetInstance"
+      },
+      "type": "sun.security.provider.SHA",
+      "methods": [
+        {
+          "name": "<init>",
+          "parameterTypes": []
+        }
+      ]
+    }
+  ],
+  "resources": [
+    {
+      "condition": {
+        "typeReached": "org.example.NativeTests"
+      },
+      "glob": "TestResource.txt"
+    }
+  ],
+  "bundles": []
+}
+----
+
+As you can see there are no more entries that contain classes from `org.junit` (as their condition) for example.
+
+
+[[maintain-generated-metadata]]
+== Maintain generated metadata
+
+If you are a library maintainer, or your application became huge, you may consider covering most of your code with tests.
+This way you can also track if your project requires updates of the existing metadata.
+Considering that you run your tests in certain CI, at some point (after updating some dependency or adding new feature/test) you may notice some test failures with missing resources errors or that something is reflectively missing...
+In that case, your metadata requires update.
+
+[WARNING]
+Be aware that if you modified existing metadata file(s) on the default location, generating a new metadata will overwrite the existing one and your manual changes will be lost.
+
+So if you modified existing metadata file(s) on the default location, please do the following:
+
+1. Move your metadata to some other, non-default location (with the `metadataCopy` task for example)
```

Comment:
:question: why is this step required if `mergeWithExisting` is true?

Reply:
Because we are executing `metadataCopy` as a separate task. When you run the agent, it will always overwrite existing metadata on the default location. By the time you want to execute `metadataCopy` (even with `mergeWithExisting` set to true) your manual changes are lost.

To avoid this:

-  you should copy metadata onto the "safe" location
- generate new metadata that will overwrite the original metadata
- then, when you execute `metadataCopy` with `mergeWithExisting` set to true, your new metadata will be merged with the old metadata placed on the "safe" location

Reply:
> When you run the agent, it will always overwrite existing metadata on the default location

It shouldn't. The agent will output the metadata in a temporary directory always. It shouldn't matter how `metadataCopy` is configured. If it's the case, then it's a bug.

Reply:
There is a misunderstanding then... When I say **default location**, in this case, I mean `temporary directory` which is the default location where agent prints metadata if you don't do anything additionally. Maybe I should assume that at this point users already moved metadata to non-temporary (or _"non-default"_) location?

Reply:
Maybe I should also add some kind of tip in the `metadataCopy` section that it is strongly recommended to move metadata from the temporary directory to standard location

Reply:
I guess it makes more sense now?

Reply:
Right, the confusion came from the fact I assumed that the metadata was already copied. I think we should recommend this indeed.